### PR TITLE
Support align by device, group query order by time desc, limit for IoTDB and TDengine and InfluxDB

### PR DESCRIPTION
**Pull request by liyuheng55555**
And refactor td's where clause addition

### CODE REVIEW

File: core/src/main/java/cn/edu/tsinghua/iot/benchmark/conf/ConfigDescriptor.java [hunk 1]
```diff
@@ -588,9 +594,25 @@ private boolean checkConfig() {
...
+    }
     return result;
   }
 
+  private boolean isGoodDB() {
+    DBType dbType = config.getDbConfig().getDB_SWITCH().getType();
+    DBVersion dbVersion = config.getDbConfig().getDB_SWITCH().getVersion();
+    return dbType.equals(DBType.IoTDB)
+        || dbVersion.equals(DBVersion.TDengine_3)
+        || dbVersion.equals(DBVersion.InfluxDB_2);
```
Comment (OneSizeFitsQuorum) [inline core/src/main/java/cn/edu/tsinghua/iot/benchmark/conf/ConfigDescriptor.java:613]: remove this line

File: core/src/main/java/cn/edu/tsinghua/iot/benchmark/conf/ConfigDescriptor.java [hunk 2]
```diff
@@ -588,9 +594,25 @@ private boolean checkConfig() {
     }
     result &= checkDeviceNumPerWrite();
     result &= checkTag();
+    if (!isGoodDB()) {
+      if (config.isALIGN_BY_DEVICE()) {
+        LOGGER.error("ALIGN_BY_DEVICE is not supported for this database");
```
Comment (OneSizeFitsQuorum) [inline core/src/main/java/cn/edu/tsinghua/iot/benchmark/conf/ConfigDescriptor.java:599]: update result

File: core/src/main/java/cn/edu/tsinghua/iot/benchmark/conf/ConfigDescriptor.java [hunk 3]
```diff
@@ -588,9 +594,25 @@ private boolean checkConfig() {
...
+      }
+      if (config.getRESULT_ROW_LIMIT() >= 0) {
+        LOGGER.error("RESULT_ROW_LIMIT is not supported for this database");
+      }
+    }
     return result;
   }
 
+  private boolean isGoodDB() {
```
Comment (OneSizeFitsQuorum) [inline core/src/main/java/cn/edu/tsinghua/iot/benchmark/conf/ConfigDescriptor.java:608]: notCommonlyUsedDB

File: configuration/conf/config.properties [hunk 4]
```diff
@@ -297,7 +297,8 @@
 # Q8  最近点查询 select time, v1... where device = ? and time = max(time)
 # Q9  倒序范围查询（只限制起止时间）select v1... from data where time > ? and time < ? and device in ? order by time desc
 # Q10 倒序带值过滤的范围查询 select v1... from data where time > ? and time < ? and v1 > ? and device in ? order by time desc
-# OPERATION_PROPORTION=1:0:0:0:0:0:0:0:0:0:0
+# Q11 分组聚合查询，倒序
+# OPERATION_PROPORTION=1:0:0:0:0:0:0:0:0:0:0:0
```
Comment (OneSizeFitsQuorum) [inline configuration/conf/config.properties:301]: 考虑兼容性

File: iotdb-1.1/src/main/java/cn/edu/tsinghua/iot/benchmark/iotdb110/IoTDBSessionBase.java [hunk 5]
```diff
@@ -169,7 +169,13 @@ public Status insertOneBatchByRecords(IBatch batch) {
...
+  protected Status addSomeClauseAndExecuteQueryAndGetStatus(String sql, Operation operation) {
+    if (config.getRESULT_ROW_LIMIT() >= 0) {
+      sql += " limit " + config.getRESULT_ROW_LIMIT();
+    }
+    if (config.isALIGN_BY_DEVICE()) {
+      sql += " align by device";
+    }
     String executeSQL;
     if (config.isIOTDB_USE_DEBUG() && random.nextDouble() < config.getIOTDB_USE_DEBUG_RATIO()) {
       executeSQL = "debug " + sql;
```
Comment (OneSizeFitsQuorum) [inline iotdb-1.1/src/main/java/cn/edu/tsinghua/iot/benchmark/iotdb110/IoTDBSessionBase.java:181]: 少 last 查询特判

File: tdengine-3.0/src/main/java/cn/edu/tsinghua/iot/benchmark/tdengine3/TDengine.java [hunk 6]
```diff
@@ -178,6 +181,7 @@ public Double registerSchema(List<DeviceSchema> schemaList) throws TsdbException
 
   private synchronized void initOnlyOnce()
       throws TsdbException, BrokenBarrierException, InterruptedException {
+    new OperationController(1).getNextOperationType();
```
Comment (OneSizeFitsQuorum) [inline tdengine-3.0/src/main/java/cn/edu/tsinghua/iot/benchmark/tdengine3/TDengine.java:184]: 删除

File: tdengine-3.0/src/main/java/cn/edu/tsinghua/iot/benchmark/tdengine3/TDengine.java [hunk 7]
```diff
@@ -443,52 +487,67 @@ private Status executeQueryAndGetStatus(String sql) {
...
-      List<DeviceSchema> devices, String sqlHeader, double valueThreshold) {
-    StringBuilder builder = new StringBuilder(sqlHeader);
-    for (Sensor sensor : devices.get(0).getSensors()) {
-      builder.append(" AND ").append(sensor.getName()).append(" > ").append(valueThreshold);
+  private static List<DeviceSchema> getTableNameFilterForAlignByDevice(RangeQuery rangeQuery) {
+    if (config.isALIGN_BY_DEVICE()) {
+      return rangeQuery.getDeviceSchema();
     }
-    return builder.toString();
+    return null;
```
Comment (OneSizeFitsQuorum) [inline tdengine-3.0/src/main/java/cn/edu/tsinghua/iot/benchmark/tdengine3/TDengine.java:543]: emptyList

File: configuration/conf/config.properties [hunk 8]
```diff
@@ -386,6 +387,14 @@
 # 查询语句的随机数种子
 # QUERY_SEED=151658
 
+# 行数限制, -1表示不使用limit语句
+# RESULT_ROW_LIMIT=-1
+
+# 是否align by device
+# 实测发现, iotdb对于align by device的使用有诸多限制, 推荐设置OPERATION_PROPORTION不超过范围1:1:1:0:0:0:0:0:0:1:0:0
+# td可设置1:1:1:1:1:1:1:1:1:1:1:0
```
Comment (OneSizeFitsQuorum) [inline configuration/conf/config.properties:395]: 目前仅支持 iotdb 和 tdengine 写一下

File: core/src/main/java/cn/edu/tsinghua/iot/benchmark/conf/ConfigDescriptor.java (hunk 1 above)
Comment (liyuheng55555) [reply core/src/main/java/cn/edu/tsinghua/iot/benchmark/conf/ConfigDescriptor.java:613]: check influxdb v1 instead

File: core/src/main/java/cn/edu/tsinghua/iot/benchmark/conf/ConfigDescriptor.java (hunk 2 above)
Comment (liyuheng55555) [reply core/src/main/java/cn/edu/tsinghua/iot/benchmark/conf/ConfigDescriptor.java:599]: updated

File: core/src/main/java/cn/edu/tsinghua/iot/benchmark/conf/ConfigDescriptor.java (hunk 3 above)
Comment (liyuheng55555) [reply core/src/main/java/cn/edu/tsinghua/iot/benchmark/conf/ConfigDescriptor.java:608]: of course

File: configuration/conf/config.properties (hunk 4 above)
Comment (liyuheng55555) [reply configuration/conf/config.properties:301]: think了

File: iotdb-1.1/src/main/java/cn/edu/tsinghua/iot/benchmark/iotdb110/IoTDBSessionBase.java (hunk 5 above)
Comment (liyuheng55555) [reply iotdb-1.1/src/main/java/cn/edu/tsinghua/iot/benchmark/iotdb110/IoTDBSessionBase.java:181]: 加了

File: tdengine-3.0/src/main/java/cn/edu/tsinghua/iot/benchmark/tdengine3/TDengine.java (hunk 7 above)
Comment (liyuheng55555) [reply tdengine-3.0/src/main/java/cn/edu/tsinghua/iot/benchmark/tdengine3/TDengine.java:543]: good point!

File: tdengine-3.0/src/main/java/cn/edu/tsinghua/iot/benchmark/tdengine3/TDengine.java (hunk 6 above)
Comment (liyuheng55555) [reply tdengine-3.0/src/main/java/cn/edu/tsinghua/iot/benchmark/tdengine3/TDengine.java:184]: shan liao

File: configuration/conf/config.properties (hunk 8 above)
Comment (liyuheng55555) [reply configuration/conf/config.properties:395]: 三处已添加

File: configuration/conf/config.properties [hunk 9]
```diff
@@ -386,6 +387,16 @@
 # 查询语句的随机数种子
 # QUERY_SEED=151658
 
+# 行数限制, -1表示不使用limit语句
+# 目前仅支持iotdb、influxdb v1、tdengine-3
+# RESULT_ROW_LIMIT=-1
+
+# 是否align by device
+# 目前仅支持iotdb、influxdb v1、tdengine-3
+# 实测发现, iotdb对于align by device的使用有诸多限制, 推荐设置OPERATION_PROPORTION不超过范围1:1:1:0:0:0:0:0:0:1:0:0
```
Comment (OneSizeFitsQuorum) [inline configuration/conf/config.properties:396]: 调整一下措辞

File: influxdb/src/main/java/cn/edu/tsinghua/iot/benchmark/influxdb/InfluxDB.java [hunk 10]
```diff
@@ -289,7 +298,11 @@ private InfluxDataModel createDataModel(
     return model;
   }
 
-  private Status executeQueryAndGetStatus(String sql) {
+  private Status addTailClausesAndExecuteQueryAndGetStatus(String sql) {
+    if (config.getRESULT_ROW_LIMIT() >= 0) {
+      sql += " limit " + config.getRESULT_ROW_LIMIT();
+    }
+    LOGGER.info("do query: {}", sql);
```
Comment (OneSizeFitsQuorum) [inline influxdb/src/main/java/cn/edu/tsinghua/iot/benchmark/influxdb/InfluxDB.java:305]: 去掉

Comment (liyuheng55555) [reply influxdb/src/main/java/cn/edu/tsinghua/iot/benchmark/influxdb/InfluxDB.java]: 去了嗷

File: configuration/conf/config.properties (hunk 9 above)
Comment (liyuheng55555) [reply configuration/conf/config.properties]: 调整了嗷

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Pull request merged**